hal/stm32f373/i2c.c

```diff
@@ -1,391 +1,401 @@
-        return;
-    sys_enter_critical_section();   // lock while changing things so an isr does not find a half setup read
-   
-    I2C_Timeout = I2C_LONG_TIMEOUT;
-    while(I2C_GetFlagStatus(i2c->channel, I2C_ISR_BUSY) != RESET)
-    {
-        if((I2C_Timeout--) == 0) return -1;
-    }
-    if (i2c->read_buf != NULL || i2c->read_count != 0)
-        ///@todo read in progress already
-        goto done;
-    //load the read info
-    i2c->read_buf = buf;
-    i2c->read_buf_len = len;
-    i2c->read_count = 0;
-    i2c->read_complete_cb = cb;
-    i2c->read_complete_param = param;    
-    // enable rx and interrupt to kick off the read
-    I2C_TransferHandling(i2c->channel, device_address, len, I2C_AutoEnd_Mode, I2C_Generate_Start_Read);  
-    /* Wait until TXIS flag is set */
-    I2C_Timeout = I2C_LONG_TIMEOUT;
-    while(I2C_GetFlagStatus(i2c->channel, I2C_ISR_RXNE) == RESET)
-    {
-        if((I2C_Timeout--) == 0) return -1;
-    }
-    if (i2c->rx_dma)
-    {
-        // dma receiver 
-        i2c->rx_dma_req.complete = i2c_rx_dma_complete;
-        i2c->rx_dma_req.complete_param = i2c;
-        i2c->rx_dma_req.dma = i2c->rx_dma;
-        i2c_dma_cfg(i2c, I2C_DMA_DIR_RX, &i2c->rx_dma_req, i2c->read_buf, i2c->read_buf_len);
-        I2C_DMACmd(i2c->channel, I2C_DMAReq_Rx, ENABLE);
-        dma_request(&i2c->rx_dma_req);
-    }
-    else
-    {
-        // interrupt receiver
-        I2C_ITConfig(i2c->channel, I2C_IT_RXI, ENABLE);
-    }
+	// sanity checks
+	if (len < 1)
+		///@todo invalid input parameters
+		return -1;
+
+	sys_enter_critical_section();   // lock while changing things so an isr does not find a half setup read
+
+	uint32_t timeout = I2C_LONG_TIMEOUT;
+	while(I2C_GetFlagStatus(i2c->channel, I2C_ISR_BUSY) != RESET)
+	{
+		if((timeout--) == 0) 
+			goto done;
+	}
+
+	if (i2c->read_buf != NULL || i2c->read_count != 0)
+		///@todo read in progress already
+		goto done;
+
+	//load the read info
+	i2c->read_buf = buf;
+	i2c->read_buf_len = len;
+	i2c->read_count = 0;
+	i2c->read_complete_cb = cb;
+	i2c->read_complete_param = param;    
+	// enable rx and interrupt to kick off the read
+	I2C_TransferHandling(i2c->channel, device_address, len, I2C_AutoEnd_Mode, I2C_Generate_Start_Read);  
+	/* Wait until TXIS flag is set */
+	timeout = I2C_LONG_TIMEOUT;
+
+	while(I2C_GetFlagStatus(i2c->channel, I2C_ISR_RXNE) == RESET)
+	{
+		if((timeout--) == 0)
+			goto done;
+	}
+
+	if (i2c->rx_dma)
+	{
+		// dma receiver 
+		i2c->rx_dma_req.complete = i2c_rx_dma_complete;
+		i2c->rx_dma_req.complete_param = i2c;
+		i2c->rx_dma_req.dma = i2c->rx_dma;
+		i2c_dma_cfg(i2c, I2C_DMA_DIR_RX, &i2c->rx_dma_req, i2c->read_buf, i2c->read_buf_len);
+		I2C_DMACmd(i2c->channel, I2C_DMAReq_Rx, ENABLE);
+		dma_request(&i2c->rx_dma_req);
+	}
+	else
+	{
+		// interrupt receiver
+		I2C_ITConfig(i2c->channel, I2C_IT_RXI, ENABLE);
+	}
 done:
-    sys_leave_critical_section();
-    return;
+	sys_leave_critical_section();
+	return 0;
```

Comment:
we don't need to do it now (this is more a note for future @othane) but we should return and error code or 0 or something ... but don't worry about it for this PR :)